ion-schema/src/isl/isl_type_reference.rs

```diff
@@ -311,3 +326,90 @@ impl IslTypeRefImpl {
         }
     }
 }
+
+/// Represents a [variably occurring type reference] that will be used by `ordered_elements` and `fields` constraints.
+///  
+/// ```ion
+/// Grammar: <VARIABLY_OCCURRING_TYPE_ARGUMENT> ::= { <OCCURS>, <CONSTRAINT>... }
+///                                      | <TYPE_ARGUMENT>
+///
+///         <OCCURS> ::= occurs: <INT>
+///            | occurs: <RANGE_INT>
+///            | occurs: optional
+///            | occurs: required
+/// ```
+///
+/// [variably occurring type reference]: https://amazon-ion.github.io/ion-schema/docs/isl-2-0/spec#variably-occurring-type-arguments
+#[derive(Debug, Clone, PartialEq)]
+pub struct IslVariablyOccurringTypeRef {
+    type_ref: IslTypeRefImpl,
+    occurs: Option<Range>,
+}
+
+impl IslVariablyOccurringTypeRef {
+    pub(crate) fn new(type_ref: IslTypeRef, occurs: Option<Range>) -> Self {
+        Self {
+            type_ref: type_ref.type_reference,
+            occurs,
+        }
+    }
+
+    pub fn occurs(&self) -> Option<Range> {
+        self.occurs.to_owned()
+    }
+
+    /// Tries to create an [IslVariablyOccurringTypeRef] from the given Element
+    pub fn from_ion_element(
+        isl_version: IslVersion,
+        value: &Element,
+        inline_imported_types: &mut Vec<IslImportType>,
+    ) -> IonSchemaResult<Self> {
+        let type_ref = IslTypeRefImpl::from_ion_element_with_occurs_flag(
+            isl_version,
+            value,
+            inline_imported_types,
+            true,
+        )?;
+        let occurs: Option<Range> = {
+            if let IslTypeRefImpl::Anonymous(isl_type, _) = &type_ref {
+                isl_type.constraints().iter().find_map(|c| match c {
+                    IslConstraintImpl::Occurs(occurs) => Some(occurs.to_owned()),
+                    _ => None,
+                })
+            } else {
+                None
+            }
+        };
+
+        Ok(IslVariablyOccurringTypeRef { type_ref, occurs })
+    }
+
+    /// Resolves an [IslVariablyOccurringTypeRef] into a [VariablyOccurringTypeRef] using the type_store
+    pub fn resolve_type_reference(
+        &self,
+        constraint_name: &str,
+        isl_version: IslVersion,
+        type_store: &mut TypeStore,
+        pending_types: &mut PendingTypes,
+    ) -> IonSchemaResult<VariablyOccurringTypeRef> {
+        let type_ref = IslTypeRefImpl::resolve_type_reference(
+            isl_version,
+            &self.type_ref,
+            type_store,
+            pending_types,
+        )?;
+
+        let occurs: Range = match &self.occurs {
+            None => {
+                if constraint_name == "fields" {
+                    Range::optional()
+                } else {
+                    Range::required()
+                }
+            }
+            Some(range) => range.to_owned(),
+        };
+
+        Ok(VariablyOccurringTypeRef::new(type_ref, occurs))
```

Comment:
This seems out of place. It's not a part of resolving the type reference. Rather it's a context-dependent part of reading the ISL in the first place.

Reply:
modified this to be part of the construction process rather then type reference resolution.